src/hmm.rs

```diff
@@ -10,47 +10,67 @@ lazy_static! {
     static ref RE_SKIP: Regex = Regex::new(r"([a-zA-Z0-9]+(?:.\d+)?%?)").unwrap();
 }
 
-pub type StatusSet = [f64; 4];
-
+pub const NUM_STATES: usize = 4;
+
+pub type StateSet = [f64; NUM_STATES];
+
+/// Result of hmm is a labeling of each Unicode Scalar Value in the input
+/// string with Begin, Middle, End, or Single. These denote the proposed
+/// segments. A segment is one of the following two patterns.
+///
+///   Begin, [Middle...], End
+///   Single
+///
+/// Each state in the enum is also assigned an index value from 0-3 that
+/// can be used as an index into an array representing data pertaining
+/// to that state.
+///
+/// WARNING: the data file format for hmm.model comments imply one can
+/// reassign the index values of each state at the top but `build.rs`
+/// currently ignores the mapping. Do not reassign these indicies without
+/// verifying hot it interacts with `build.rs`
 #[derive(Debug, PartialEq, Eq, Hash, PartialOrd, Ord, Clone, Copy)]
-pub enum Status {
-    B = 0,
-    E = 1,
-    M = 2,
-    S = 3,
+pub enum State {
+    Begin = 0,
+    End = 1,
+    Middle = 2,
+    Single = 3,
 }
 
-static PREV_STATUS: [[Status; 2]; 4] = [
-    [Status::E, Status::S], // B
-    [Status::B, Status::M], // E
-    [Status::M, Status::B], // M
-    [Status::S, Status::E], // S
-];
+// Mapping representing the allow transitiongs into the given state.
+static ALLOWED_PREV_STATUS: [[State; 2]; NUM_STATES] = {
+    let mut valid_transitions_from: [[State; 2]; NUM_STATES] = [[State::Begin, State::Begin]; NUM_STATES];
+    valid_transitions_from[State::Begin as usize] = [State::End, State::Single];
+    valid_transitions_from[State::End as usize] = [State::Begin, State::Middle];
+    valid_transitions_from[State::Middle as usize] = [State::Middle, State::Begin];
+    valid_transitions_from[State::Single as usize] = [State::Single, State::End];
+    valid_transitions_from
```

Comment:
I'm finding this `as usize` harder to read, how about leave it as-is (change the variant names only).